src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,446 @@
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var typeName = WellKnownTypeNames.SystemRuntimeInteropServicesRuntimeInformation;
+
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(typeName + "Helper", out var runtimeInformationType))
+                {
+                    runtimeInformationType = context.Compilation.GetOrCreateTypeByMetadataName(typeName);
+                }
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesOSPlatform, out var osPlatformType))
+                {
+                    return;
+                }
+
+                var guardMethods = GetRuntimePlatformGuardMethods(runtimeInformationType!);
+
+                context.RegisterOperationBlockStartAction(context => AnalyzerOperationBlock(context, guardMethods, osPlatformType));
+            });
+
+            static ImmutableArray<IMethodSymbol> GetRuntimePlatformGuardMethods(INamedTypeSymbol runtimeInformationType)
+            {
+                return runtimeInformationType.GetMembers().OfType<IMethodSymbol>().Where(m =>
+                    s_platformCheckMethodNames.Contains(m.Name) && !m.Parameters.IsEmpty).ToImmutableArray();
+            }
+        }
+
+        private void AnalyzerOperationBlock(OperationBlockStartAnalysisContext context, ImmutableArray<IMethodSymbol> guardMethods, INamedTypeSymbol osPlatformType)
+        {
+            var platformSpecificOperations = PooledDictionary<IOperation, ImmutableArray<PlatformAttributeInfo>>.GetInstance();
+
+            context.RegisterOperationAction(context =>
+            {
+                AnalyzeInvocationOperation(context.Operation, context, ref platformSpecificOperations);
+            }
+            , OperationKind.DelegateCreation, OperationKind.EventReference, OperationKind.FieldReference,
+            OperationKind.Invocation, OperationKind.ObjectCreation, OperationKind.PropertyReference);
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                try
+                {
+                    if (platformSpecificOperations.Count == 0)
+                    {
+                        return;
+                    }
+
+                    if (guardMethods.IsEmpty || !(context.OperationBlocks.GetControlFlowGraph() is { } cfg))
+                    {
+                        ReportDiagnosticsForAll(platformSpecificOperations, context);
+                        return;
+                    }
+
+                    var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(context.Compilation);
+                    var analysisResult = GlobalFlowStateAnalysis.TryGetOrComputeResult(
+                        cfg, context.OwningSymbol, CreateOperationVisitor, wellKnownTypeProvider,
+                        context.Options, MinimumOsRule, performValueContentAnalysis: true,
```

Comment:
> performValueContentAnalysis: true

We should not always force execute value content analysis. It is expensive and should only be used for cases that really require it, which should be minimal as most callers will not save result of platform guard method invocations into a local, but instead have it directly as the if condition. Can we restore the functionality from existing analyzer that determines when this is required?

Reply:
I am actually talked about this before and decided to set true, the existing logic for determining it was not so feasible with this implementation, but i will see how to make it work